wandb/wandb
# feat(artifacts): keep uncommitted uploads in separate staging area

Part of [WB-11386](https://wandb.atlassian.net/browse/WB-11386)

Note that this does _not_ fix [WB-11386](https://wandb.atlassian.net/browse/WB-11386), but it enables a simpler and more general solution.

Description
-----------
Use a separate staging area for artifact files that haven't been committed (instead of relying on the cache).

See design doc, [stage artifact uploads](https://www.notion.so/wandbai/Stage-artifact-uploads-44fabbaea7af4094807d6116482b70ab)

**Other changes**
- We need to put the staging data somewhere, so I'm using `appdirs` to default to a platform-compliant data directory. Since I added this I also used it for the cache directory. On linux, this doesn't change its location, but it moves it to the "platform preferred" directory on macOS and Windows.

Testing
-------
New tests:
- Add multiple copies of the same file concurrently and ensure that all copies get cleaned up correctly
- Clear the cache while an upload is pending and confirm that the artifact commits successfully
- Verify that a file that's been uploaded gets copied to the cache afterwards
- Edit a file immediately after adding it (before upload) and verify the file committed is the original
- Check that files are cleared out of the staging area once they've been committed

Checklist
-------
- [x] Include reference to internal ticket "Fixes WB-NNNN" and/or GitHub issue "Fixes #NNNN" (if applicable)
- [x] Ensure PR title compliance with the [conventional commits standards](https://github.com/wandb/wandb/blob/master/CONTRIBUTING.md#conventional-commits)
